### PR TITLE
[flux.2 LoRA] make lora training compatible with flux.2 klein kv

### PR DESCRIPTION
**Pull request by linoytsaban**
adds a flux.2 klein kv training script, with additional general changes that can be propagated to other lora scripts as well: 

- [x] **Default aspect ratio buckets** - uses preset aspect ratio buckets to avoid needing to pass them manually. 

- [x] **`--caption_dropout_rate`** - Randomly replaces prompts with empty strings at the given rate (default 0.05). Forces the model to learn from visual signal alone on some steps to improve robustness.

- [x] **`--shift_timesteps`** - resolution-adaptive timestep sampling. Samples from sigmoid distribution then warps with `t' = (t·μ)/(1+(μ-1)·t)` where μ scales with latent sequence length. higher resolution gets more high-noise training. default behaviour in popular trainers like ai-toolkit and khoya.

- [x] **Cache keying fix** - latent/prompt caches bug fixed.

- [ ] **Multiple conditions support** - allows for multiple image conditions per example.

### CODE REVIEW

File: examples/dreambooth/train_dreambooth_lora_flux2_klein_kv_img2img.py [hunk 1]
```diff
@@ -0,0 +1,2046 @@
+#!/usr/bin/env python
+# coding=utf-8
+# Copyright 2025 The HuggingFace Inc. team. All rights reserved.
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_flux2_klein_kv_img2img.py]:
```suggestion
# Copyright 2026 The HuggingFace Inc. team. All rights reserved.
```

File: examples/dreambooth/train_dreambooth_lora_flux2_klein_kv_img2img.py [hunk 2]
```diff
@@ -0,0 +1,2046 @@
...
+                }
+            )
+
+    del pipeline
+    free_memory()
+
+    return images
+
+
+def module_filter_fn(mod: torch.nn.Module, fqn: str):
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_flux2_klein_kv_img2img.py:241]: Maybe for a different PR: we could move it to `training_utils.py` and name it `module_filter_fn_torchao`?

File: examples/dreambooth/train_dreambooth_lora_flux2_klein_kv_img2img.py [hunk 3]
```diff
@@ -0,0 +1,2046 @@
...
+    batch["indices"] = [example["index"] for example in examples]  # propagate indices for cache keying
+    if any("cond_images" in example for example in examples):
+        cond_pixel_values = [example["cond_images"] for example in examples]
+        cond_pixel_values = torch.stack(cond_pixel_values)
+        cond_pixel_values = cond_pixel_values.to(memory_format=torch.contiguous_format).float()
+        batch.update({"cond_pixel_values": cond_pixel_values})
+    return batch
+
+
+class BucketBatchSampler(BatchSampler):
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_flux2_klein_kv_img2img.py:1031]: Can probably be moved to `training_utils.py`?

File: examples/dreambooth/train_dreambooth_lora_flux2_klein_kv_img2img.py [hunk 4]
```diff
@@ -0,0 +1,2046 @@
...
+    if args.gradient_checkpointing:
+        transformer.enable_gradient_checkpointing()
+
+    if args.lora_layers is not None:
+        target_modules = [layer.strip() for layer in args.lora_layers.split(",")]
+    else:
+        # train transformer_blocks and single_transformer_blocks
+        target_modules = ["to_k", "to_q", "to_v", "to_out.0"] + [
+            "to_qkv_mlp_proj",
+            *[f"single_transformer_blocks.{i}.attn.to_out" for i in range(24)],
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_flux2_klein_kv_img2img.py:1267]: ⚡️

File: examples/dreambooth/train_dreambooth_lora_flux2_klein_kv_img2img.py [hunk 5]
```diff
@@ -0,0 +1,2046 @@
...
+    # pack the statically computed variables appropriately here. This is so that we don't
+    # have to pass them to the dataloader.
+    if not train_dataset.custom_instance_prompts:
+        prompt_embeds = instance_prompt_hidden_states
+        text_ids = instance_text_ids
+
+    # if cache_latents is set to True, we encode images to latents and store them.
+    # Similar to pre-encoding in the case of a single instance prompt, if custom prompts are provided
+    # we encode them in advance as well.
+    precompute_latents = args.cache_latents or train_dataset.custom_instance_prompts
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_flux2_klein_kv_img2img.py:1531]: What was the fix needed to ensure the keying fix for precomputed latents?

File: examples/dreambooth/train_dreambooth_lora_flux2_klein_kv_img2img.py [hunk 6]
```diff
@@ -0,0 +1,2046 @@
...
+        schedule_timesteps = noise_scheduler_copy.timesteps.to(accelerator.device)
+        timesteps = timesteps.to(accelerator.device)
+        step_indices = [(schedule_timesteps == t).nonzero().item() for t in timesteps]
+
+        sigma = sigmas[step_indices].flatten()
+        while len(sigma.shape) < n_dim:
+            sigma = sigma.unsqueeze(-1)
+        return sigma
+
+    def calculate_shift(
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_flux2_klein_kv_img2img.py:1684]: Could use it from the pipeline itself?

File: examples/dreambooth/train_dreambooth_lora_flux2_klein_kv_img2img.py [hunk 7]
```diff
@@ -0,0 +1,2046 @@
...
+        m = (max_shift - base_shift) / (max_seq_len - base_seq_len)
+        b = base_shift - m * base_seq_len
+        mu = image_seq_len * m + b
+        return mu
+
+    # Pre-compute empty prompt embeddings for caption dropout.
+    # When caption dropout triggers, we replace the real prompt embedding with this.
+    # Note: empty_prompt_embeds and empty_text_ids are computed above before the text encoder is freed.
+    if args.caption_dropout_rate > 0.0:
+        if empty_prompt_embeds is None:
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_flux2_klein_kv_img2img.py:1717]: This will probably not be the case since we're already doing:

```py
if args.caption_dropout_rate > 0.0:
        logger.info("Pre-computing empty prompt embeddings for caption dropout...")
        with offload_models(text_encoding_pipeline, device=accelerator.device, offload=args.offload):
            empty_prompt_embeds, empty_text_ids = compute_text_embeddings("", text_encoding_pipeline)
```

File: examples/dreambooth/train_dreambooth_lora_flux2_klein_kv_img2img.py [hunk 8]
```diff
@@ -0,0 +1,2046 @@
...
+
+            with accelerator.accumulate(models_to_accumulate):
+                if train_dataset.custom_instance_prompts:
+                    # Clone when caption dropout is active to avoid mutating the cache.
+                    if args.caption_dropout_rate > 0.0:
+                        prompt_embeds = prompt_embeds_cache[cache_key].clone()
+                        text_ids = text_ids_cache[cache_key].clone()
+                    else:
+                        prompt_embeds = prompt_embeds_cache[cache_key]
+                        text_ids = text_ids_cache[cache_key]
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_flux2_klein_kv_img2img.py:1740]: Why is this needed?

File: examples/dreambooth/train_dreambooth_lora_flux2_klein_kv_img2img.py (hunk 2 above)
Comment (linoytsaban) [reply examples/dreambooth/train_dreambooth_lora_flux2_klein_kv_img2img.py:241]: maybe do that in a seprate refactor PR since this also persists in other lora scripts?

File: examples/dreambooth/train_dreambooth_lora_flux2_klein_kv_img2img.py (hunk 3 above)
Comment (linoytsaban) [reply examples/dreambooth/train_dreambooth_lora_flux2_klein_kv_img2img.py:1031]: maybe do that in a seprate refactor PR since this also persists in other lora scripts?

**Comment by linoytsaban** [comment]
closing this PR as these changes will be applied in a new PR to existing script for non-distilled models
